File: go/internal/signer/local_test.go
package signer_test

import (
	"context"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"testing"

	gethcrypto "github.com/ethereum/go-ethereum/crypto"

	"github.com/rootwarp/eth-utils/go/internal/signer"
	internaltx "github.com/rootwarp/eth-utils/go/internal/tx"
)

// validHexKey generates a fresh valid private key hex string for tests.
func validHexKey(t *testing.T) (string, string) {
	t.Helper()
	priv, err := gethcrypto.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey: %v", err)
	}
	keyHex := hex.EncodeToString(gethcrypto.FromECDSA(priv))
	addrHex := gethcrypto.PubkeyToAddress(priv.PublicKey).Hex()
	return keyHex, addrHex
}

func holeskyUnsignedTx() internaltx.UnsignedTx {
	return internaltx.UnsignedTx{
		ChainID:              17000,
		To:                   "0x4242424242424242424242424242424242424242",
		Value:                "0x1BC16D674EC80000", // 2 ETH in wei
		Data:                 "0xabcd",
		Gas:                  21000,
		MaxFeePerGas:         "0x3B9ACA00",  // 1 gwei
		MaxPriorityFeePerGas: "0x3B9ACA00",  // 1 gwei
		Nonce:                0,
		Type:                 "0x2",
	}
}

func TestNewLocalSignerFromHex_Valid(t *testing.T) {
	keyHex, wantAddr := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	signed, err := s.Sign(context.Background(), holeskyUnsignedTx())
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if !strings.EqualFold(signed.From, wantAddr) {
		t.Errorf("From = %q, want %q", signed.From, wantAddr)
	}
}

func TestNewLocalSignerFromHex_ValidWithPrefix(t *testing.T) {
	keyHex, _ := validHexKey(t)
	_, err := signer.NewLocalSignerFromHex("0x" + keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex with 0x prefix: %v", err)
	}
}

func TestNewLocalSignerFromHex_InvalidLength(t *testing.T) {
	cases := []struct {
		name  string
		input string
	}{
		{"too_short", "ab"},
		{"63_hex_chars", "0x" + strings.Repeat("a", 63)},
		{"65_hex_chars", "0x" + strings.Repeat("a", 65)},
		{"empty", ""},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			_, err := signer.NewLocalSignerFromHex(tc.input)
			if !errors.Is(err, signer.ErrInvalidKey) {
				t.Errorf("want ErrInvalidKey, got %v", err)
			}
		})
	}
}

func TestNewLocalSignerFromHex_BadHex(t *testing.T) {
	badInput := "0x" + strings.Repeat("z", 64) // 'z' is not valid hex
	_, err := signer.NewLocalSignerFromHex(badInput)
	if !errors.Is(err, signer.ErrInvalidKey) {
		t.Errorf("want ErrInvalidKey for bad hex, got %v", err)
	}
}

func TestNewLocalSignerFromHex_ZeroScalar(t *testing.T) {
	zeroKey := strings.Repeat("0", 64)
	_, err := signer.NewLocalSignerFromHex(zeroKey)
	if !errors.Is(err, signer.ErrInvalidKey) {
		t.Errorf("want ErrInvalidKey for zero scalar, got %v", err)
	}
}

func TestNewLocalSignerFromHex_ErrorDoesNotIncludeKey(t *testing.T) {
	badInput := "0x" + strings.Repeat("f", 63) // wrong length, but memorable bytes
	_, err := signer.NewLocalSignerFromHex(badInput)
	if err == nil {
		t.Fatal("expected error")
	}
	// The error must not leak the input bytes.
	if strings.Contains(err.Error(), strings.Repeat("f", 63)) {
		t.Error("error message contains key material")
	}
}

func TestNewLocalSignerFromEnv_Set(t *testing.T) {
	keyHex, _ := validHexKey(t)
	t.Setenv("TEST_LOCAL_SIGNER_KEY", keyHex)
	s, err := signer.NewLocalSignerFromEnv("TEST_LOCAL_SIGNER_KEY")
	if err != nil {
		t.Fatalf("NewLocalSignerFromEnv: %v", err)
	}
	defer s.Close()

	_, err = s.Sign(context.Background(), holeskyUnsignedTx())
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
}

func TestNewLocalSignerFromEnv_Missing(t *testing.T) {
	t.Setenv("TEST_MISSING_KEY", "")
	_, err := signer.NewLocalSignerFromEnv("TEST_MISSING_KEY")
	if err == nil {
		t.Fatal("expected error for empty env var")
	}
	// Error must reference the var name but not contain key material.
	if !strings.Contains(err.Error(), "TEST_MISSING_KEY") {
		t.Errorf("error should mention env var name, got: %v", err)
	}
}

func TestLocalSigner_Sign_RoundTrip(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	unsigned := holeskyUnsignedTx()
	signed, err := s.Sign(context.Background(), unsigned)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}

	// Decode the RawRLP back.
	rawHex := strings.TrimPrefix(signed.RawRLP, "0x")
	raw, err := hex.DecodeString(rawHex)
	if err != nil {
		t.Fatalf("hex.DecodeString RawRLP: %v", err)
	}

	// Verify the raw bytes decode without error via RoundTrip check.
	if len(raw) < 2 {
		t.Fatalf("RawRLP too short: %d bytes", len(raw))
	}
	// Type-2 transactions have 0x02 prefix byte.
	if raw[0] != 0x02 {
		t.Errorf("RawRLP[0] = 0x%02x, want 0x02 (EIP-2718 type-2)", raw[0])
	}

	// Basic field checks.
	if signed.Hash == "" || !strings.HasPrefix(signed.Hash, "0x") {
		t.Errorf("Hash = %q, want 0x-prefixed", signed.Hash)
	}
	if signed.R == "" || !strings.HasPrefix(signed.R, "0x") {
		t.Errorf("R = %q, want 0x-prefixed", signed.R)
	}
	if signed.S == "" || !strings.HasPrefix(signed.S, "0x") {
		t.Errorf("S = %q, want 0x-prefixed", signed.S)
	}
	if signed.V != "0" && signed.V != "1" {
		t.Errorf("V = %q, want decimal 0 or 1", signed.V)
	}
	if !strings.HasPrefix(signed.From, "0x") {
		t.Errorf("From = %q, want 0x-prefixed", signed.From)
	}
	if signed.Unsigned.ChainID != unsigned.ChainID {
		t.Errorf("Unsigned.ChainID = %d, want %d", signed.Unsigned.ChainID, unsigned.ChainID)
	}
}

func TestLocalSigner_Sign_SenderRecovery(t *testing.T) {
	priv, err := gethcrypto.GenerateKey()
	if err != nil {
		t.Fatalf("GenerateKey: %v", err)
	}
	keyHex := hex.EncodeToString(gethcrypto.FromECDSA(priv))
	wantAddr := gethcrypto.PubkeyToAddress(priv.PublicKey).Hex()

	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	signed, err := s.Sign(context.Background(), holeskyUnsignedTx())
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}

	if !strings.EqualFold(signed.From, wantAddr) {
		t.Errorf("From = %q, want %q", signed.From, wantAddr)
	}
}

func TestLocalSigner_Sign_ChainID17000(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	unsigned := holeskyUnsignedTx()
	signed, err := s.Sign(context.Background(), unsigned)
	if err != nil {
		t.Fatalf("Sign: %v", err)
	}
	if signed.Unsigned.ChainID != 17000 {
		t.Errorf("ChainID = %d, want 17000", signed.Unsigned.ChainID)
	}
}

func TestLocalSigner_Sign_Cancelled(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	ctx, cancel := context.WithCancel(context.Background())
	cancel() // pre-cancel

	_, err = s.Sign(ctx, holeskyUnsignedTx())
	if err == nil {
		t.Fatal("expected error for cancelled context")
	}
	if !errors.Is(err, context.Canceled) {
		t.Errorf("want context.Canceled, got %v", err)
	}
}

func TestLocalSigner_Close_Idempotent(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	if err := s.Close(); err != nil {
		t.Errorf("first Close: %v", err)
	}
	if err := s.Close(); err != nil {
		t.Errorf("second Close: %v", err)
	}
}

func TestLocalSigner_Sign_AfterClose(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	if err := s.Close(); err != nil {
		t.Fatalf("Close: %v", err)
	}
	_, err = s.Sign(context.Background(), holeskyUnsignedTx())
	if !errors.Is(err, signer.ErrSignerClosed) {
		t.Errorf("want ErrSignerClosed, got %v", err)
	}
}

func TestLocalSigner_Name(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	if s.Name() != "local" {
		t.Errorf("Name() = %q, want %q", s.Name(), "local")
	}
}

func TestLocalSigner_RequiresUserInteraction(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	if s.RequiresUserInteraction() {
		t.Error("RequiresUserInteraction() = true, want false")
	}
}

func TestLocalSigner_Address_ReturnsKeyAddress(t *testing.T) {
	keyHex, wantAddr := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()

	addr, err := s.Address()
	if err != nil {
		t.Fatalf("Address: %v", err)
	}
	if !strings.EqualFold(addr.Hex(), wantAddr) {
		t.Errorf("Address() = %q, want %q", addr.Hex(), wantAddr)
	}
}

func TestLocalSigner_Address_AfterClose(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	if err := s.Close(); err != nil {
		t.Fatalf("Close: %v", err)
	}
	if _, err := s.Address(); !errors.Is(err, signer.ErrSignerClosed) {
		t.Errorf("want ErrSignerClosed, got %v", err)
	}
}

// --- Must Fix 1: ChainID=0 rejection ---

func TestLocalSigner_Sign_ChainID0_Rejected(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	unsigned := holeskyUnsignedTx()
	unsigned.ChainID = 0
	_, err = s.Sign(context.Background(), unsigned)
	if !errors.Is(err, signer.ErrInvalidChainID) {
		t.Fatalf("want ErrInvalidChainID, got %v", err)
	}
}

// --- Must Fix 2: Empty/invalid gas field rejection ---

func TestLocalSigner_Sign_EmptyMaxFeePerGas_Rejected(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	unsigned := holeskyUnsignedTx()
	unsigned.MaxFeePerGas = ""
	_, err = s.Sign(context.Background(), unsigned)
	if err == nil {
		t.Fatal("expected error for empty MaxFeePerGas")
	}
}

func TestLocalSigner_Sign_EmptyMaxPriorityFeePerGas_Rejected(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	unsigned := holeskyUnsignedTx()
	unsigned.MaxPriorityFeePerGas = ""
	_, err = s.Sign(context.Background(), unsigned)
	if err == nil {
		t.Fatal("expected error for empty MaxPriorityFeePerGas")
	}
}

func TestLocalSigner_Sign_InvalidMaxFeeHex_Rejected(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	unsigned := holeskyUnsignedTx()
	unsigned.MaxFeePerGas = "0xgg"
	_, err = s.Sign(context.Background(), unsigned)
	if err == nil {
		t.Fatal("expected error for invalid MaxFeePerGas hex")
	}
}

func TestLocalSigner_Sign_InvalidMaxPriorityFeeHex_Rejected(t *testing.T) {
	keyHex, _ := validHexKey(t)
	s, err := signer.NewLocalSignerFromHex(keyHex)
	if err != nil {
		t.Fatalf("NewLocalSignerFromHex: %v", err)
	}
	defer s.Close()
	unsigned := holeskyUnsignedTx()
	unsigned.MaxPriorityFeePerGas = "0xgg"
	_, err = s.Sign(context.Background(), unsigned)
	if err == nil {
		t.Fatal("expected error for invalid MaxPriorityFeePerGas hex")
	}
}

// TestLocalSigner_Sign_VariousChainIDs verifies signing works for mainnet and other chains.
func TestLocalSigner_Sign_VariousChainIDs(t *testing.T) {
	keyHex, _ := validHexKey(t)
	for _, chainID := range []uint64{1, 5, 11155111, 17000} {
		t.Run(fmt.Sprintf("chainID_%d", chainID), func(t *testing.T) {
			s, err := signer.NewLocalSignerFromHex(keyHex)
			if err != nil {
				t.Fatalf("NewLocalSignerFromHex: %v", err)
			}
			defer s.Close()

			unsigned := holeskyUnsignedTx()
			unsigned.ChainID = chainID
			signed, err := s.Sign(context.Background(), unsigned)
			if err != nil {
				t.Fatalf("Sign chainID=%d: %v", chainID, err)
			}
			if signed.V != "0" && signed.V != "1" {
				t.Errorf("V = %q, want decimal 0 or 1", signed.V)
			}
		})
	}
}

File: rust/crates/signer/tests/local_test.rs
//! Black-box tests for the local signer, ported from
//! `go/internal/signer/local_test.go`.
//!
//! Go tests generated a fresh random key per test (`gethcrypto.GenerateKey`);
//! the port uses fixed deterministic keys so the crate needs no RNG
//! dependency. The expected address is derived through the public
//! `LocalSigner::address()` accessor instead of geth's `PubkeyToAddress`.

use eth_deposit_signer::{
    new_local_signer_from_env, new_local_signer_from_hex, Signer, SignerError,
};
use eth_deposit_tx::UnsignedTx;

/// A valid deterministic 32-byte secp256k1 scalar (Go: validHexKey).
const VALID_KEY_HEX: &str = "0202020202020202020202020202020202020202020202020202020202020202";

fn holesky_unsigned_tx() -> UnsignedTx {
    UnsignedTx {
        chain_id: 17000,
        to: "0x4242424242424242424242424242424242424242".into(),
        value: "0x1BC16D674EC80000".into(), // 2 ETH in wei
        data: "0xabcd".into(),
        gas: 21000,
        max_fee_per_gas: "0x3B9ACA00".into(),          // 1 gwei
        max_priority_fee_per_gas: "0x3B9ACA00".into(), // 1 gwei
        nonce: 0,
        tx_type: "0x2".into(),
        ..UnsignedTx::default()
    }
}

fn addr_hex(addr: &[u8; 20]) -> String {
    format!("0x{}", hex::encode(addr))
}

// Go: TestNewLocalSignerFromHex_Valid
#[test]
fn new_local_signer_from_hex_valid() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).expect("NewLocalSignerFromHex");
    let want_addr = addr_hex(&s.address().unwrap());

    let signed = s.sign(&holesky_unsigned_tx()).expect("Sign");
    assert!(
        signed.from.eq_ignore_ascii_case(&want_addr),
        "From = {:?}, want {:?}",
        signed.from,
        want_addr
    );
    let _ = s.close();
}

// Go: TestNewLocalSignerFromHex_ValidWithPrefix
#[test]
fn new_local_signer_from_hex_valid_with_prefix() {
    new_local_signer_from_hex(&format!("0x{VALID_KEY_HEX}"))
        .expect("NewLocalSignerFromHex with 0x prefix");
}

// Go: TestNewLocalSignerFromHex_InvalidLength
#[test]
fn new_local_signer_from_hex_invalid_length() {
    let cases: &[(&str, String)] = &[
        ("too_short", "ab".to_string()),
        ("63_hex_chars", format!("0x{}", "a".repeat(63))),
        ("65_hex_chars", format!("0x{}", "a".repeat(65))),
        ("empty", String::new()),
    ];
    for (name, input) in cases {
        let err = new_local_signer_from_hex(input).unwrap_err();
        assert!(
            matches!(err.sentinel(), SignerError::InvalidKey),
            "{name}: want InvalidKey, got {err}"
        );
    }
}

// Go: TestNewLocalSignerFromHex_BadHex
#[test]
fn new_local_signer_from_hex_bad_hex() {
    let bad_input = format!("0x{}", "z".repeat(64)); // 'z' is not valid hex
    let err = new_local_signer_from_hex(&bad_input).unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::InvalidKey),
        "want InvalidKey for bad hex, got {err}"
    );
}

// Go: TestNewLocalSignerFromHex_ZeroScalar
#[test]
fn new_local_signer_from_hex_zero_scalar() {
    let zero_key = "0".repeat(64);
    let err = new_local_signer_from_hex(&zero_key).unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::InvalidKey),
        "want InvalidKey for zero scalar, got {err}"
    );
}

// Go: TestNewLocalSignerFromHex_ErrorDoesNotIncludeKey
#[test]
fn new_local_signer_from_hex_error_does_not_include_key() {
    let bad_input = format!("0x{}", "f".repeat(63)); // wrong length, memorable bytes
    let err = new_local_signer_from_hex(&bad_input).unwrap_err();
    // The error must not leak the input bytes.
    assert!(
        !err.to_string().contains(&"f".repeat(63)),
        "error message contains key material"
    );
}

// Go: TestNewLocalSignerFromEnv_Set
#[test]
fn new_local_signer_from_env_set() {
    std::env::set_var("TEST_LOCAL_SIGNER_KEY_RS", VALID_KEY_HEX);
    let s = new_local_signer_from_env("TEST_LOCAL_SIGNER_KEY_RS").expect("NewLocalSignerFromEnv");
    s.sign(&holesky_unsigned_tx()).expect("Sign");
    let _ = s.close();
    std::env::remove_var("TEST_LOCAL_SIGNER_KEY_RS");
}

// Go: TestNewLocalSignerFromEnv_Missing
#[test]
fn new_local_signer_from_env_missing() {
    std::env::remove_var("TEST_MISSING_KEY_RS");
    let err = new_local_signer_from_env("TEST_MISSING_KEY_RS").unwrap_err();
    // Error must reference the var name but not contain key material.
    assert!(
        err.to_string().contains("TEST_MISSING_KEY_RS"),
        "error should mention env var name, got: {err}"
    );
    assert_eq!(
        err.to_string(),
        "environment variable \"TEST_MISSING_KEY_RS\" is not set or empty: invalid private key"
    );
}

// Go: TestLocalSigner_Sign_RoundTrip
#[test]
fn local_signer_sign_round_trip() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let unsigned = holesky_unsigned_tx();
    let signed = s.sign(&unsigned).expect("Sign");

    // Decode the RawRLP back.
    let raw_hex = signed.raw_rlp.strip_prefix("0x").expect("0x prefix");
    let raw = hex::decode(raw_hex).expect("hex decode RawRLP");
    assert!(raw.len() >= 2, "RawRLP too short: {} bytes", raw.len());
    // Type-2 transactions have 0x02 prefix byte.
    assert_eq!(raw[0], 0x02, "RawRLP[0] want 0x02 (EIP-2718 type-2)");

    // Basic field checks.
    assert!(signed.hash.starts_with("0x") && signed.hash.len() > 2);
    assert!(signed.r.starts_with("0x") && signed.r.len() > 2);
    assert!(signed.s.starts_with("0x") && signed.s.len() > 2);
    assert!(signed.v == "0" || signed.v == "1", "V = {:?}", signed.v);
    assert!(signed.from.starts_with("0x"));
    assert_eq!(signed.unsigned.chain_id, unsigned.chain_id);
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_SenderRecovery
#[test]
fn local_signer_sign_sender_recovery() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let want_addr = addr_hex(&s.address().unwrap());

    let signed = s.sign(&holesky_unsigned_tx()).expect("Sign");
    assert!(
        signed.from.eq_ignore_ascii_case(&want_addr),
        "From = {:?}, want {:?}",
        signed.from,
        want_addr
    );
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_ChainID17000
#[test]
fn local_signer_sign_chain_id_17000() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let signed = s.sign(&holesky_unsigned_tx()).expect("Sign");
    assert_eq!(signed.unsigned.chain_id, 17000);
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_Cancelled
#[test]
fn local_signer_sign_cancelled() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    // Pre-cancelled: the closure reports cancellation immediately.
    let err = s
        .sign_with_cancel(&holesky_unsigned_tx(), &|| true)
        .unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::Cancelled),
        "want Cancelled, got {err}"
    );
    let _ = s.close();
}

// Go: TestLocalSigner_Close_Idempotent
#[test]
fn local_signer_close_idempotent() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    assert!(s.close().is_ok(), "first close");
    assert!(s.close().is_ok(), "second close");
}

// Go: TestLocalSigner_Sign_AfterClose
#[test]
fn local_signer_sign_after_close() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    s.close().unwrap();
    let err = s.sign(&holesky_unsigned_tx()).unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::SignerClosed),
        "want SignerClosed, got {err}"
    );
}

// Go: TestLocalSigner_Name
#[test]
fn local_signer_name() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    assert_eq!(s.name(), "local");
    let _ = s.close();
}

// Go: TestLocalSigner_RequiresUserInteraction
#[test]
fn local_signer_requires_user_interaction() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    assert!(!s.requires_user_interaction());
    let _ = s.close();
}

// Go: TestLocalSigner_Address_ReturnsKeyAddress
#[test]
fn local_signer_address_returns_key_address() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    // Address must be consistent with the sender recovered from a signature.
    let addr = addr_hex(&s.address().expect("Address"));
    let signed = s.sign(&holesky_unsigned_tx()).unwrap();
    assert!(
        signed.from.eq_ignore_ascii_case(&addr),
        "Address() = {:?}, recovered From = {:?}",
        addr,
        signed.from
    );
    let _ = s.close();
}

// Go: TestLocalSigner_Address_AfterClose
#[test]
fn local_signer_address_after_close() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    s.close().unwrap();
    let err = s.address().unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::SignerClosed),
        "want SignerClosed, got {err}"
    );
}

// Go: TestLocalSigner_Sign_ChainID0_Rejected (Must Fix 1)
#[test]
fn local_signer_sign_chain_id_0_rejected() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let mut unsigned = holesky_unsigned_tx();
    unsigned.chain_id = 0;
    let err = s.sign(&unsigned).unwrap_err();
    assert!(
        matches!(err.sentinel(), SignerError::InvalidChainId),
        "want InvalidChainId, got {err}"
    );
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_EmptyMaxFeePerGas_Rejected (Must Fix 2)
#[test]
fn local_signer_sign_empty_max_fee_per_gas_rejected() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let mut unsigned = holesky_unsigned_tx();
    unsigned.max_fee_per_gas = String::new();
    assert!(s.sign(&unsigned).is_err());
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_EmptyMaxPriorityFeePerGas_Rejected
#[test]
fn local_signer_sign_empty_max_priority_fee_per_gas_rejected() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let mut unsigned = holesky_unsigned_tx();
    unsigned.max_priority_fee_per_gas = String::new();
    assert!(s.sign(&unsigned).is_err());
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_InvalidMaxFeeHex_Rejected
#[test]
fn local_signer_sign_invalid_max_fee_hex_rejected() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let mut unsigned = holesky_unsigned_tx();
    unsigned.max_fee_per_gas = "0xgg".into();
    assert!(s.sign(&unsigned).is_err());
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_InvalidMaxPriorityFeeHex_Rejected
#[test]
fn local_signer_sign_invalid_max_priority_fee_hex_rejected() {
    let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
    let mut unsigned = holesky_unsigned_tx();
    unsigned.max_priority_fee_per_gas = "0xgg".into();
    assert!(s.sign(&unsigned).is_err());
    let _ = s.close();
}

// Go: TestLocalSigner_Sign_VariousChainIDs
#[test]
fn local_signer_sign_various_chain_ids() {
    for chain_id in [1u64, 5, 11155111, 17000] {
        let s = new_local_signer_from_hex(VALID_KEY_HEX).unwrap();
        let mut unsigned = holesky_unsigned_tx();
        unsigned.chain_id = chain_id;
        let signed = s
            .sign(&unsigned)
            .unwrap_or_else(|e| panic!("Sign chainID={chain_id}: {e}"));
        assert!(
            signed.v == "0" || signed.v == "1",
            "chainID={chain_id}: V = {:?}, want decimal 0 or 1",
            signed.v
        );
        let _ = s.close();
    }
}
